Allow attaching to existing installation

nicolaferraro opened the pull request:
This fixes https://github.com/syndesisio/syndesis/issues/3092 and also allows to upgrade from a legacy installation of syndesis (a installation without operator).

It is sufficient to add the operator to the namespace and if there's a legacy installation, a Syndesis CR is created and its configuration is reversed from environment variables used during template installation.

If the operator has a different syndesis version, the upgrade is triggered automatically after the CR is merged and attached to all existing resources. I've added a state diagram to better understand what happens.

Note, creation of Syndesis resource for a legacy installation happens only at startup of the operator (and it is retried in case of error), the namespace is not watched continuously after the first check.

Here's the log of a legacy installation first attached to a Syndesis CR, then upgraded:
```
﻿time="2018-07-20T08:50:30Z" level=info msg="Go Version: go1.10"
time="2018-07-20T08:50:30Z" level=info msg="Go OS/Arch: linux/amd64"
time="2018-07-20T08:50:30Z" level=info msg="operator-sdk Version: 0.0.5+git"
time="2018-07-20T08:50:30Z" level=info msg="Watching syndesis.io/v1alpha1, Syndesis, test, 5"
time="2018-07-20T08:50:30Z" level=info msg="A legacy Syndesis installations is present in the test namespace"
time="2018-07-20T08:50:30Z" level=info msg="Creating a new Syndesis resource from legacy installation in the test namespace"
time="2018-07-20T08:50:30Z" level=info msg="Syndesis legacy installations check completed"
time="2018-07-20T08:50:30Z" level=info msg="Merging Syndesis legacy configuration into resource syndesis-app"
time="2018-07-20T08:50:30Z" level=info msg="Syndesis resource syndesis-app merged with legacy installation"
time="2018-07-20T08:50:30Z" level=info msg="Attaching Syndesis installation to resource syndesis-app"
time="2018-07-20T08:50:37Z" level=info msg="Syndesis installation attached to resource syndesis-app"
time="2018-07-20T08:50:37Z" level=info msg="Syndesis resource syndesis-app installed successfully"
time="2018-07-20T08:50:37Z" level=info msg="Starting upgrade of Syndesis resource syndesis-app from version latest to version 1.4.12"
time="2018-07-20T08:50:38Z" level=info msg="Upgrading syndesis resource syndesis-app from version latest to 1.4.12"
time="2018-07-20T08:50:40Z" level=info msg="Syndesis resource syndesis-app is currently being upgraded to version 1.4.12"
time="2018-07-20T08:50:45Z" level=info msg="Syndesis resource syndesis-app is currently being upgraded to version 1.4.12"
...
time="2018-07-20T08:52:50Z" level=info msg="Syndesis resource syndesis-app is currently being upgraded to version 1.4.12"
time="2018-07-20T08:52:55Z" level=info msg="Syndesis resource syndesis-app upgraded to version 1.4.12"
time="2018-07-20T08:52:55Z" level=info msg="Attaching Syndesis installation to resource syndesis-app"
time="2018-07-20T08:53:02Z" level=info msg="Syndesis installation attached to resource syndesis-app"
time="2018-07-20T08:53:02Z" level=info msg="Syndesis resource syndesis-app installed successfully"
```

The merged resource looks like the following at the end of the process:
```
apiVersion: syndesis.io/v1alpha1
kind: Syndesis
metadata:
  clusterName: ""
  creationTimestamp: 2018-07-20T08:50:30Z
  generation: 0
  name: syndesis-app
  namespace: test
  resourceVersion: "942934"
  selfLink: /apis/syndesis.io/v1alpha1/namespaces/test/syndesises/syndesis-app
  uid: f4f1376d-8bf9-11e8-ba07-8e1028189acc
spec:
  components:
    db:
      database: syndesis
      imageStreamNamespace: openshift
      resources:
        limits:
          memory: 255Mi
        volumeCapacity: 1Gi
      user: syndesis
    meta:
      resources: {}
    prometheus:
      resources:
        limits:
          memory: 512Mi
        volumeCapacity: 1Gi
    server:
      resources: {}
  demoData: false
  deployIntegrations: true
  registry: docker.io
  routeHostname: syndesis.192.168.42.153.nip.io
status:
  installationStatus: Installed
  version: 1.4.12
```

Configuration has been copied.

The same process of attaching the resources (but not merging configuration properties) is used after the upgrade (passing the owner references to the upgrade pod is more difficult, because there are multiple fields to be passed, this strategy is easier and needed also during the "attachment" to a legacy installation).

nicolaferraro committed:
Adding merging and attaching states

nicolaferraro committed:
Added upgrade from legacy installation

nicolaferraro committed:
Added state diagram documentation

nicolaferraro committed:
Update embedded resources

nicolaferraro committed:
Fix setting resource limits

rhuss commented:
give me 'till tomorrow, will review it then.

rhuss commented on pkg/syndesis/action/merge.go:
```diff
@@ -0,0 +1,37 @@
+
+
+func (a *Merge) CanExecute(syndesis *v1alpha1.Syndesis) bool {
+	return syndesisInstallationStatusIs(syndesis, v1alpha1.SyndesisInstallationStatusMerging)
```
Do we really need the 'Merging' status ?  mean, the legacyController could already populate the created Syndesis resource with the values from the configmap when it creates it.

This would simplify the state diagram a bit, but maybe I missed something why it would be good to decouple these two steps ?

rhuss commented on pkg/syndesis/action/attach.go:
```diff
@@ -0,0 +1,161 @@
+	// Checking that there's only one installation to avoid stealing resources
+	if anotherInstallation, err := isAnotherActiveInstallationPresent(syndesis); err != nil {
+		return err
+	} else if anotherInstallation {
```
minor: the `else` is a bit confusing here (as it evaluates the variable just set in if as side effect.

Maybe just omit the `else` ?

rhuss commented on pkg/syndesis/action/attach.go:
```diff
@@ -0,0 +1,161 @@
+
+		target.Status.InstallationStatus = v1alpha1.SyndesisInstallationStatusNotInstalled
+		target.Status.Reason = v1alpha1.SyndesisStatusReasonDuplicate
+		target.Status.Description = "Cannot merge because another Syndesis resources is present in the same namespace"
```
"attach" instead of "merge" ?

rhuss commented on docs/syndesis-operator-states.xml:
```diff
@@ -0,0 +1 @@
+<mxfile userAgent="Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36" version="8.9.7" editor="www.draw.io" type="device"><diagram id="8ecc18c7-d5c6-3091-a06c-258d5145b2f6" name="Page-1">7V1Zs6JGFP4193EsemN5nLmzJFVJKpVbqSSPqO2SQdogzr3m16eRbrAXAVnUifgwI03TIOc75zsb3Cf0vHn7koTb1c9sTqMn6MzfntDHJwgBhi7/Lxs55CMeAvnAMlnPxaRy4GX9LxWDjhjdr+d0p0xMGYvS9 ...
```
There is a connection "Attaching" --> "Not Installed" missing in case of a parallel installation.

rhuss commented on pkg/syndesis/action/attach.go:
```diff
@@ -0,0 +1,161 @@
+
+	for _, selector := range getAllManagerSelectors() {
+		for _, metaType := range getAllManagedResourceTypes() {
+
```
wouldn't it be better to iterate over all the objects stored in the template ? It looks a bit dangerous and maintenance intense to hardcode the objects contained in the template here.

Up to now we only know about the Route in the template. BTW, what do you think about removing the Syndesis `Route` object from the template, and just create it as an extra object, either directly from the go-code or as an included yml ? That way we would avoid that dance with instantiating the template twice, fishing out the Route object etc. would be much directer. Even when we want to backport that to our manual installation, it would be easy to achieve and we could benefit from the automatically set route. I.e. we should handle the route on the same footstep as the service account as we need generated values from both (token, hostName) to instantiate the template.

nicolaferraro replied:
I didn't iterate over the template to avoid problems for things that for some reasons are removed from the template or change their name starting from some version. I may take "object types" from the template and leave room for adding other types if needed.. that would be less error prone..

About removing the Route from the template, that would simplify a lot the code as you've noticed.

rhuss commented on pkg/syndesis/action/attach.go:
```diff
@@ -0,0 +1,161 @@
+func getAllManagerSelectors() []metav1.ListOptions {
+	return []metav1.ListOptions {
+		{
+			LabelSelector: "syndesis.io/app=syndesis,syndesis.io/type=infrastructure",
```
actually I would thought that this selector should be enough. Do you know which other objects which neede to be attached requires special label ?

IMO it should be enough to select on "app=syndesis", and then on any object for doing that kind of attachment.

'happy to fix that on the template side.

nicolaferraro replied:
For the TODO, the syndesis.io/app is different and there's also a error with the DB PVC not belonging to the "infrastructure" type. I've fixed the template but I expect existing installations not to be fixed...

rhuss commented on pkg/syndesis/action/attach.go:
```diff
@@ -0,0 +1,161 @@
+
+
+func (a *Attach) CanExecute(syndesis *v1alpha1.Syndesis) bool {
+	return syndesisInstallationStatusIs(syndesis, v1alpha1.SyndesisInstallationStatusAttaching)
```
In which circumstance beside after an upgrade could it be that we lose the owner reference ? If its only after the upgrade, wouldn't it be more direct to just update the reference on all objects of any Kind with the label "app=syndesis" (see also my comment below about the labels) ?

nicolaferraro replied:
I've experienced issues on getting resources of "Any" kind with the client.. Will retry, but the Kind seems mandatory.

rhuss commented on pkg/syndesis/action/attach.go:
```diff
@@ -0,0 +1,161 @@
+
+			options := sdk.WithListOptions(&selector)
+			list := metav1.List{
+				TypeMeta: metaType,
```
Is a type required for the list operation or is there a way to list all objects regardless of the kind which carry a certain label ?

nicolaferraro replied:
As said, will retry, but seems not possible...

rhuss commented on pkg/syndesis/action/initialize.go:
```diff
@@ -34,8 +38,14 @@ func (a *Initialize) Execute(syndesis *v1alpha1.Syndesis) error {
 		logrus.Error("Cannot initialize Syndesis resource ", syndesis.Name, ": duplicate")
+	} else if syndesisAlreadyInstalled {
+		// One uninitialized Syndesis CR and Syndesis already installed: attaching the installation to the resource
+		target.Status.InstallationStatus = v1alpha1.SyndesisInstallationStatusAttaching
```
In what case can this happen ? Having a single fresh Syndesis resource with no state but with an already installed Syndesis ? This could be only the legacy use case, or ? And that is tackled by the legacyController during startup which creates a "Merge" phased resource. Only when in the meantime someone deploys a fresh Syndesis resource, a kind of race could happen.

I don't think we should put too much effort on the legacy case, as this will be a one shot thing. And we should remove that handling for 7.2 imo where we should require, when updating from 7.0 to 7.2 it needs to be done via 7.1

nicolaferraro replied:
Yes, removing it..

rhuss commented on pkg/syndesis/action/upgrade.go:
```diff
@@ -149,7 +149,7 @@ func (a *Upgrade) Execute(syndesis *v1alpha1.Syndesis) error {
 func upgradeCompleted(syndesis *v1alpha1.Syndesis, newVersion string) error {
 	target := syndesis.DeepCopy()
-	target.Status.InstallationStatus = v1alpha1.SyndesisInstallationStatusInstalled
+	target.Status.InstallationStatus = v1alpha1.SyndesisInstallationStatusAttaching
```
tbh I would do the owner attachment directly after an successful upgrade and during the creation of the synthetic Syndesis resource for legacy installtation.

rhuss reviewed (changes requested):
Thanks a lot ! Especially for the state diagram and doc update 👍 

Tbh, I question a bit the introduction of the `Merge` and `Attaching` status (btw, we should change that to 'phase' as it seems that's how its called in k8s land).

They introduce an extra complexity in the state model which I don't think is necessary:

* For updating legacy installation (which will exists only for 7.1, we should remove it for 7.2) we can do the creation of a "Installed" syndesis, already attached, in the legacy controller directly.
* For re-attaching after an upgrade this can also be performed directly after a successful upgrade in the "Upgrading" phase.

I don't see any other use case where reattaching could be required.

IMO it's always a kind of trade of between flexibility (by moving fine granular steps for handling async) and simplicity (simplifying the state diagram, with more hard wired steps).

Its always a balancing act, but in this case (and especially it involves a temporary functionality, so that at the end reattaching is only required after an upgrade) I tend to the simplicity.

wdyt ?

rhuss commented:
> (passing the owner references to the upgrade pod is more difficult, because there are multiple fields to be passed, this strategy is easier and needed also during the "attachment" to a legacy installation).

btw, I don't meant to provide this to the upgrade pod, but let the upgrade pod doing its work and then do the attachment directly in upgrade.go (at the spot where you set the "Attaching" status)

nicolaferraro commented:
I'll update the PR later. The only thing I'd like to keep is the "Attaching" state, just because it's better to do changes on other resources in the critical region of the sdk loop, instead of doing changes in another goroutine.

rhuss commented:
> The only thing I'd like to keep is the "Attaching" state, just because it's better to do changes on other resources in the critical region of the sdk loop, instead of doing changes in another goroutine.

Actually, I didn't meant to do it in an extra gorouting  but just synchronously [after the upgrade](https://github.com/syndesisio/syndesis-operator/blob/4b77372379f31ff72059290ee7c1893be3a0e215/pkg/syndesis/action/upgrade.go#L118).

Also it will be only after the upgrade where this is required (as updating legacy installation will exist only for a limited amount of time).

Not sure that I understood the _it's better to do changes on other resources in the critical region of the sdk loop_, though :)

nicolaferraro commented:
Yes, sorry, the upgrade from legacy installation is currently done partially by a different goroutine that hands the work over to the sdk loop in the "Merging" stage (or in the "Aggregating" stage if "Merging" is removed). So it's better to keep a state for the upgrade from legacy... And I though to use it also after upgrade..
Btw, we can also call that state "LegacyUpgrade" and use it only for legacy. While we can do the "attaching" in upgrade.go as you said..

rhuss commented:
Yeah, would like to make the legacy stuff easily removable because for 7.1 we will remove that as for 7.0 users to upgrade we will force them to go over 7.1.

nicolaferraro committed:
Updated docs

nicolaferraro committed:
Implemented new state diagram

nicolaferraro committed:
Renaming installationStatus into phase

nicolaferraro committed:
Updating documentation

nicolaferraro committed:
Extract types from template and fix json spec

nicolaferraro commented:
I should have addressed most of the issues.. including using "phase" instead of "installationStatus". Not moved the route out of the template. I think we can better do it when we move to monorepo.

rhuss commented on pkg/syndesis/action/upgradelegacy.go:
```diff
@@ -0,0 +1,66 @@
+	for _, that := range lst.Items {
+		if that.Name != syndesis.Name &&
+			that.Status.Phase != v1alpha1.SyndesisPhaseNotInstalled &&
+			that.Status.Phase != v1alpha1.SyndesisStatusReasonMissing {
```
Does this compile for you ? (phase != reason)

rhuss replied:
Guess you mean `SyndesisPhaseMissing` ?

nicolaferraro replied:
Good catch.. don't know why it compiles..

rhuss commented on resources/template.yaml:
```diff
@@ -544,6 +544,7 @@ objects:
     labels:
       app: syndesis
       syndesis.io/app: syndesis
+      syndesis.io/type: infrastructure
```
I'll take this over to syndesis, too ...

rhuss commented on pkg/syndesis/operation/attach.go:
```diff
@@ -0,0 +1,96 @@
+			LabelSelector: "app=syndesis,syndesis.io/app=todo",
+		},
+		{
+			LabelSelector: "syndesis.io/app=syndesis,syndesis.io/component=syndesis-db",
```
going to fix this in syndesis, too (and the template), so maybe add a comment that this can be removed soon ?

rhuss commented on pkg/syndesis/operation/attach.go:
```diff
@@ -0,0 +1,96 @@
+	// Using a map to remove duplicates
+	metas := make(map[string]metav1.TypeMeta, 0)
+	for _, kType := range templTypes {
+		addResourceTypes(metas, kType.APIVersion, kType.Kind)
```
I think you could use `kType` directly as a key in a map, avoiding to create the key yourself.

AFAIK the only requirement for a key is that it needs to be comparable using `==`. `metav1.TypeMeta` should be comparable as it is a struct with two strings only, and a struct comparison boils down to the comparison of its elements.

rhuss replied:
So you could use a `map[metav1.TypeMeta]bool` as a Set. Actually, you could even change `template.GetDeclaredResourceTypes` to return that set or already a unique slice.

But that should not hinder us to merge that PR now. 'happy to pick this up to find out whether we could use TypeMeta as key.

nicolaferraro replied:
I assumed that complex types as map keys were forbidden in Go.. bu I'll let you investigate :)

rhuss reviewed (changes requested):
Thanks a lot, that's really awesome ! Happy to merge when the one single bug below is fixed (phase vs reason in the check).

rhuss commented on pkg/syndesis/operation/attach.go:
```diff
@@ -0,0 +1,96 @@
+
+func addMissingResourceTypes(types map[string]metav1.TypeMeta) {
+	// Add here any resource type that should be attached but may not be present in the template
+	addResourceTypes(types, "batch/v1beta1", "CronJob")
```
A CronJob resource included in the latest template btw.

nicolaferraro committed:
Fix wrong constant in comparison

nicolaferraro commented:
Done!

rhuss reviewed (approved):
🕺
